viewing_party/party.py

```diff
@@ -1,23 +1,229 @@
+    """
+    if not user_data["watched"]:
+        return None
+    genre_count = {} # dictionary to store genre and counts.
+    highest_count = 0
+    most_common_genre = ""
+    for movie in user_data["watched"]:
+        if movie["genre"] in genre_count:
+            genre_count[movie["genre"]] += 1
+            if genre_count[movie["genre"]] > highest_count:
+                highest_count = genre_count[movie["genre"]]
+                most_common_genre = movie["genre"]
+        else: 
+            genre_count[movie["genre"]] = 1
+    return most_common_genre 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+# TC: O(n*f*m)
+# n -- length of user['watched']
+# f -- number of friends
+# m -- average number of movies watched by each friend
+# SC: O(k) length of unique movies watched by user, worst case O(n)
+def get_unique_watched(user_data): 
+    """
+    Get the movies which the user has watched but their friends have not.
+    Parameter: user_data(dict)
+    Return: List of user's unique movies(list)
+    """
+    """
+    # older version:
+    copy_of_user_data_watched = list(user_data['watched'])
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie in copy_of_user_data_watched:
+                copy_of_user_data_watched.remove(movie)
+    return copy_of_user_data_watched
+    """
+    #updated code 
+    # edge cases: empty watched list for user, or no friends;
+    if not user_data["watched"]:
+        return []
+    if not user_data["friends"]:
+        return user_data["watched"]
+    user_unique_watched = []
+    for movie in user_data["watched"]:
+        is_unique = True        
+        for friend in user_data["friends"]:
+            if not friend["watched"]:
+                continue
+            if movie in friend["watched"]:
+                is_unique = False
+                break
+        if is_unique:
+            user_unique_watched.append(movie) 
+    return user_unique_watched
+
+# TC: O(f * m * (n + k))
+# SC: O(k)
+# f -- number of friends
+# m -- average number of movies watched by each friend
+# n -- length of user['watched']
+# k -- length of friends_unque_watched_movie
+
+def get_friends_unique_watched(user_data):
+    """
+    Get the movies which the friends have watched, but user has not.
+    Parameter: user_data(dict)
+    Return: List of friends' unique movies(list)
+    """
+    friends_unique_watched = []
+    if not user_data["friends"]:
+        return []
+    
+    for friend in user_data["friends"]:
+        if not friend["watched"]:
+            continue
+        for movie in friend["watched"]:
+            if not user_data["watched"] or \
+                (movie not in user_data["watched"] and 
+                 movie not in friends_unique_watched):
+                friends_unique_watched.append(movie)
+    return friends_unique_watched

```

Comment:
The same idea as the previous function applies here! You could separate everything out into different sets and find their difference! If you want to get even fancier, you could parse out what it similar between these two functions and create a couple helper functions that can convert the user's "watched" list into a set of titles and the friends "watched" list into a different set of titles as well!